### PR TITLE
Add selectors to useContext calls

### PR DESCRIPTION
**Pull request by cvolzke4**
## Summary

Add selector functions to useContext calls (using useContextSelectorShallow).

See also: https://docs.google.com/document/d/18m9HamHkOvvTrv-mSOJawha0K4KTdn49HrnfXShULDI/edit#heading=h.21vtkbcpsflf

This will allow specific attributes of the context state to be selected without re-rendering after every context state change.

React profiler of scrolling through media after this series of changes:
https://share.getcloudapp.com/YEu14qDg

## To-do

Next steps:
- Add useMemo at the end of each context provider.
- Use Pure Components for simple components.
- Prevent action functions from depending on state changes.

---

Fixes #1466

### CODE REVIEW

File: assets/src/edit-story/components/autoSaveHandler/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/autoSaveHandler/index.js]: These eslints will get very old very soon. Any way to deal with this automatically?

/cc @swissspidy for any ideas? Doesn't necessarily have to block this PR, but would be very good to see if we can resolve.

Alternative to this would be to extract this function into a top-level `const storySelector = ({...}) => {}` function. Would that help with lints? If so, I think that could have some cons and pros. But I feel like i'd rather keep it inline if we can do it well.

File: assets/src/edit-story/components/inspector/document/status.js
Comment (dvoytenko) [inline assets/src/edit-story/components/inspector/document/status.js]: Eslint makes things really hard here, but I'm thinking if we addressed that if the following form would work well:

```
  const { status, password, updateStory, deleteStory } = useStory(({
      state: { story: { status, password } },
      actions: { updateStory, deleteStory },
  }) => ({status, password, updateStory, deleteStory}));
```

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: assets/src/edit-story/components/inspector/document/status.js
Comment (cvolzke4) [reply assets/src/edit-story/components/inspector/document/status.js]: Much nicer :)
Updated.

File: assets/src/edit-story/components/canvas/canvasProvider.js
Comment (cvolzke4) [inline assets/src/edit-story/components/canvas/canvasProvider.js]: Not sure why the linter prefers this syntax when the return value object is wrapped.

File: assets/src/edit-story/utils/context.js
Comment (swissspidy) [inline assets/src/edit-story/utils/context.js]: Unnecessary import? If it's for the docblock one could do `@param {import('react').Context}` or `/** @typedef {import('react').Context} context Context */` and then `@param {Context} context Context`

File: assets/src/edit-story/components/autoSaveHandler/index.js
Comment (swissspidy) [reply assets/src/edit-story/components/autoSaveHandler/index.js]: Agreed. We should come up with a solution to improve this.

The rule itself is valuable to prevent headaches, so I'd suggest finding a workaround.

In other areas I see usage like this, and it works well especially if only a few items are selected:

```js
const { status, saveStory } = useStory((state) => ({
  status: state.state.status,
  saveStory: state.actions.saveStory,
}));
```

Alternatively, maybe our `objectPick` utility could be useful?

```js
const {
  currentPage,
  selectedElementIds,
  toggleElementInSelection,
  setSelectedElementsById,
} = useStory((state) => ({
  ...objectPick(state.state, ['currentPage', 'selectedElementIds']),
  ...objectPick(state.actions, ['toggleElementInSelection', 'setSelectedElementsById'])
}));
```

For comparison, that's half the lines of what's currently in `assets/src/edit-story/components/canvas/canvasProvider.js`:

```js
const {
  currentPage,
  selectedElementIds,
  toggleElementInSelection,
  setSelectedElementsById,
} = useStory(
  ({
    // eslint-disable-next-line no-shadow
    state: { currentPage, selectedElementIds },
    // eslint-disable-next-line no-shadow
    actions: { toggleElementInSelection, setSelectedElementsById },
  }) => {
    return {
      currentPage,
      selectedElementIds,
      toggleElementInSelection,
      setSelectedElementsById,
    };
  }
);
```

File: assets/src/edit-story/components/canvas/useCanvasKeys.js
Comment (swissspidy) [inline assets/src/edit-story/components/canvas/useCanvasKeys.js]: FYI, you can just do `eslint-disable no-shadow` somewhere and `eslint-enable no-shadow` afterwards instead of repeating this 7 times.

**Comment by barklund** [comment]
Did you try to run any speed tests or profiling on this before/after? I don't really have any suggestions to how though. Just to see if there's a measurable gain here. I like it, but only if the benefits are significant.

File: assets/src/edit-story/components/autoSaveHandler/index.js
Comment (dvoytenko) [reply assets/src/edit-story/components/autoSaveHandler/index.js]: Not a big fun of `objectPick` tbh. Here's my thinking here: the shadowing here happens in this form:

```
const {prop1, prop2} = functionToProvideProps(({prop1, prop2}) => {...})
```

I.e. an inline call shadows the top-level props it itself produces. IMHO this is a pretty justified case for shadowing. So I'm wondering if there might be some settings/customization available to avoid it. I'd even argue that there's no true prop shadowing here with `const` because until the `functionToProvideProps` returns the top-level `prop1` and `prop2` are actually not defined.

**Comment by dvoytenko** [comment]
> Did you try to run any speed tests or profiling on this before/after? I don't really have any suggestions to how though. Just to see if there's a measurable gain here. I like it, but only if the benefits are significant.

@barklund, I think the description provides a doc and a profiling data. And from what it looks like, the performance gain is significant. But where I'm sold on this, is a tighter isolation of possible performance issues from spilling to the app at large. Let us know what you think.

File: assets/src/edit-story/components/autoSaveHandler/index.js
Comment (dvoytenko) [reply assets/src/edit-story/components/autoSaveHandler/index.js]: @swissspidy wdyt if we tried `hoist: never` here? Not sure it fits 100%, but sounds like it might.

**Review by dvoytenko — APPROVED** [review]
I'm LGTM on this but would really like to find some solution for no-shadow if we can before merging. Please also wait for approval from at least one more person (@swissspidy and/or @barklund) before merging.

File: assets/src/edit-story/utils/context.js
Comment (cvolzke4) [reply assets/src/edit-story/utils/context.js]: Updated.

**Comment by cvolzke4** [comment]
> Did you try to run any speed tests or profiling on this before/after? I don't really have any suggestions to how though. Just to see if there's a measurable gain here. I like it, but only if the benefits are significant.

See: https://docs.google.com/document/d/18m9HamHkOvvTrv-mSOJawha0K4KTdn49HrnfXShULDI/edit#heading=h.4bti0lhcgu9m

File: assets/src/edit-story/components/canvas/useCanvasKeys.js
Comment (cvolzke4) [reply assets/src/edit-story/components/canvas/useCanvasKeys.js]: Updated, for selectors that have more than 2 lines of disabled lint rules.

File: assets/src/edit-story/components/autoSaveHandler/index.js
Comment (swissspidy) [reply assets/src/edit-story/components/autoSaveHandler/index.js]: Unfortunately changing the `hoist` option doesn't do anything, just tried.

To get to an agreement I propose the following:

1. Turn the rule's severity from `error` to `warn` for the time being
1. Report this to ESLint and maybe even contribute a fix (https://github.com/eslint/eslint/issues/12687 might cover this already)
1. Once fixed in ESLint, change severity back to `error`

Comment (barklund) [reply assets/src/edit-story/components/autoSaveHandler/index.js]: This would work for me. And yes, I definitely think that https://github.com/eslint/eslint/issues/12687 solves the issue.

File: assets/src/edit-story/components/canvas/displayLayer.js
Comment (barklund) [inline assets/src/edit-story/components/canvas/displayLayer.js:48]: I'm adding this as a comment here, even though it doesn't apply to this selector in particular:

Would it make sense to over time develop more specific hooks for common selections as shortcuts?

So this one would be just:

```js
  const { currentPage } = useCurrentPage();
```

But other occasions might be a little less specific like:

```js
  const { updateElementsByResourceId } = useStoryAction('updateElementsByResourceId');
```

Also, is there any reason that we return an object with variables as keys here? Is that required by the underlying `useSelector` helper?

I.e. why couldn't this original code here not just be:

```js
  const currentPage = useStory((state) => state.state.currentPage);
```

And thus the suggested helpers be:

```js
  const currentPage = useCurrentPage();
  const updateElementsByResourceId = useStoryAction('updateElementsByResourceId');
```

**Review by barklund — APPROVED** [review]
Everything works great and this seems like a great addition and improvement to the codebase.

I will not only gladly approve of it here, but in general recommend it when implementing the context provider pattern (as I've been touting its virtues elsewhere on many occasions).

I do want the eslint false negative issue solved first, but @swissspidy's [suggestion](https://github.com/google/web-stories-wp/pull/1705#discussion_r426541027) of temporarily changing to a warning while waiting for a proper fix, is good with me.

Finally I have a single comment/suggestion below.

**Review by barklund — CHANGES_REQUESTED** [review]
Revisiting this, editmode appears to be broken in this PR. Not sure exactly why or how (no errors thrown), but when you click to edit a text element, the element simply disappears. When you click "outside" or press escape, the element reappears.

selector_editmode_broken

**Comment by cvolzke4** [comment]
> Revisiting this, editmode appears to be broken in this PR. Not sure exactly why or how (no errors thrown), but when you click to edit a text element, the element simply disappears. When you click "outside" or press escape, the element reappears.
> 
> selector_editmode_broken

This is because of a recursion problem in useUploader, which is addressed here:
https://github.com/google/web-stories-wp/pull/1799

File: assets/src/edit-story/components/canvas/displayLayer.js
Comment (cvolzke4) [reply assets/src/edit-story/components/canvas/displayLayer.js:48]: > const { currentPage } = useCurrentPage();

I'm not sure about this one, as it obfuscates the context the attribute is queried from, and require a lot of them (eg. usePages, useIsSaving + 10 others).

> const { updateElementsByResourceId } = useStoryAction('updateElementsByResourceId');

This may be more useful, it's easy to misspell the attribute name in the string but it's also easy to pick that up in code review (and, javascript). This version is more type safe if typescript were introduced later:
> const currentPage = useStory((state) => state.state.currentPage);

I can see a couple of downsides to this:

1. If more than one attribute is needed, it's more work to refactor the call to useXX to add the additional state or actions, without making multiple calls to 'useContextSelector' which is probably unnecessary. For example, refactoring this:

```js
const currentPage = useStory((state) => state.state.currentPage);
```
to

```js
const { selectedElements, stylePresets, updateStory } = useStory(
  ({
    state: { selectedElements, story: { stylePresets } },
    actions: { updateStory },
  }) => ({ selectedElements, stylePresets, updateStory })
);
```
Most developers will be lazy and do this instead:

```js
const currentPage = useStory((state) => state.state.currentPage);
const stylePresets = useStory((state) => state.state.story.stylePresets);
const updateStory = useStory((state) => state.actions.updateStory);
```
which means 3 calls to useContextSelector.

2. This isn't so important, but useStory() performs a shallow comparison selector. This means all of the attributes of 'currentPage' are compared instead of only the reference to currentPage, which may be unnecessary. Of the order of microseconds not ms, so this may not matter very much.

3. Some state/actions have more nesting, like 'state.story.stylePresets'.

In a separate PR, if everyone likes this approach I can change the above useStory() example (as an example) to this:

```js
const currentPage = useStory('currentPage');
const stylePresets = useStory((state) => state.state.story.stylePresets);
const updateStory = useStoryAction('updateStory');
```
This disregards caveats 2 and 3 above :D but perhaps this is a readability improvement and is worth it.

WDYT?

File: assets/src/edit-story/components/autoSaveHandler/index.js
Comment (cvolzke4) [reply assets/src/edit-story/components/autoSaveHandler/index.js]: Updated to warn.

File: assets/src/edit-story/components/canvas/displayLayer.js
Comment (barklund) [reply assets/src/edit-story/components/canvas/displayLayer.js:48]: Let's work on the API in a separate PR. Several of your points are excellent and we need to strike the right balance between UX and DX.

**Review by barklund — COMMENTED** [review]
I'm a little hesitant with approving this one, as it changes a lot of stuff in the application before we have good integration test coverage. We'd definitely need to apply full QA process before merging this to validate, that **everything** works as before.

So, no merging before full QA has been performed [according to QA spec](https://docs.google.com/document/d/1oZXjSJX30JLsIgCqOaER0Uff3Y_Tr-NCiQIfrYUnX9Y/edit?ts=5ebebd11#heading=h.cttk1x7k1ujn).

**Comment by dvoytenko** [comment]
> I'm a little hesitant with approving this one, as it changes a lot of stuff in the application before we have good integration test coverage. We'd definitely need to apply full QA process before merging this to validate, that **everything** works as before.
> 
> So, no merging before full QA has been performed [according to QA spec](https://docs.google.com/document/d/1oZXjSJX30JLsIgCqOaER0Uff3Y_Tr-NCiQIfrYUnX9Y/edit?ts=5ebebd11#heading=h.cttk1x7k1ujn).

@barklund @cvolzke4 That's reasonable. Can we prioritize this for manual QA ASAP?

File: assets/src/edit-story/components/inspector/design/useDesignPanels.js
Comment (dvoytenko) [inline assets/src/edit-story/components/inspector/design/useDesignPanels.js:41]: nit: inline

File: assets/src/edit-story/components/inspector/document/publish/index.js
Comment (dvoytenko) [inline assets/src/edit-story/components/inspector/document/publish/index.js:72]: nit: inline

File: assets/src/edit-story/components/library/panes/media/mediaPane.js
Comment (dvoytenko) [inline assets/src/edit-story/components/library/panes/media/mediaPane.js:153]: nit: inline

File: assets/src/edit-story/components/panels/layer/useLayerSelection.js
Comment (dvoytenko) [inline assets/src/edit-story/components/panels/layer/useLayerSelection.js:41]: nit: inline

File: assets/src/edit-story/components/panels/stylePreset/panel.js
Comment (dvoytenko) [inline assets/src/edit-story/components/panels/stylePreset/panel.js:55]: nit: inline

File: assets/src/edit-story/components/inspector/design/useDesignPanels.js
Comment (cvolzke4) [reply assets/src/edit-story/components/inspector/design/useDesignPanels.js:41]: Strangely, this causes the argument list to be formatted on one line:

  } = useStory(
    ({ state: { selectedElementIds, selectedElements }, actions: { deleteSelectedElements, updateElementsById } }) => ({
      selectedElementIds,
      selectedElements,
      deleteSelectedElements,
      updateElementsById,
    });
  );

File: assets/src/edit-story/components/inspector/document/publish/index.js
Comment (cvolzke4) [reply assets/src/edit-story/components/inspector/document/publish/index.js:72]: As above.

File: assets/src/edit-story/components/library/panes/media/mediaPane.js
Comment (cvolzke4) [reply assets/src/edit-story/components/library/panes/media/mediaPane.js:153]: As above.

File: assets/src/edit-story/components/panels/layer/useLayerSelection.js
Comment (cvolzke4) [reply assets/src/edit-story/components/panels/layer/useLayerSelection.js:41]: As above.

File: assets/src/edit-story/components/panels/stylePreset/panel.js
Comment (cvolzke4) [reply assets/src/edit-story/components/panels/stylePreset/panel.js:55]: As above.

**Comment by cvolzke4** [comment]
> Revisiting this, editmode appears to be broken in this PR. Not sure exactly why or how (no errors thrown), but when you click to edit a text element, the element simply disappears. When you click "outside" or press escape, the element reappears.

Actually this was a problem with a selector that renames an element:

const { editingElement: editingElementId } = useCanvas((state) => ({
  editingElementId: state.state.editingElementId,
}));

This has been fixed, and I've re-checked each diff once again for good measure.

**Comment by cvolzke4** [comment]
Testing notes
-------------
Please reach out if there are any details that need elaborating, or if a face to face will help.

This change affects the way data is consumed throughout the story editor. This includes all aspects of the story editor including media gallery, upload, canvas, property inspection, etc.

This change does not affect dashboard functionality.

During development, the kinds of defects that were found were diverse:
- Story editor does not load.
- Double-click of text element causes the element to disappear.
- Infinite recursion during story editor load (pre-existing bug surfaced in a different and more breaking way).

I don't anticipate any issues, but due to the broad nature of the change a full regression of the Story Editor may be justified.

This will likely mean finding random bugs unrelated to this change. It may be helpful to open a browser with this change in one tab, and a version without this change in another tab, so that the source of any defect can be verified.

**Review by dvoytenko — APPROVED** [review]

**Comment by dvoytenko** [comment]
@cvolzke4 @barklund I wanted to see if we should expedite and get this in this week so we still have some time to pick up the bugs next week? Did QA do a pass on this?

**Comment by dvoytenko** [comment]
Looks like this is Ok'd by QA.

**Review by barklund — APPROVED** [review]
LGTM

**Pull request merged**
